### PR TITLE
Implement determineShift. Close #16

### LINKED ISSUE #16: Implement general rotation algorithm
Depends on #9 and #15.

In order to implement general rotation algorithm, we need to combine existing helper methods.
First implementation will not care about the position of tetromino relative to edges of the field and blocks on the field (no problems when coordinates are less than zero or more than size of the field or when rotation results overlaps blocks already placed on the field).

- Create some type of tetromino graphically in some rotational phase (buildTetromino/getCoordinates)
- Determine top-left corner of tetromino (calculatorCoordinateOfTopLeft)
- Rotate tetromino counter-clockwise around the top-left corner (rotatesArrayOfCoordinates)
- Determine left-right and top-bottom shift amount (determineShift)
- Shift tetromino coordinates this amount (shiftCoordinates)

## determineShift(typeOfTetromino, rotationPhaseOfTetromino)

In order to determine shift amount, you need additional helper function:
Parameters: 
- typeOfTetromino - I, O, Z, S, J, L or T (1)
- rotationPhaseOfTetromino - single value representing current phase of tetromino. Dor example I-tetromino can be "vertical" or "horizontal" (2)

This function should return (x,y)-object representing by how much each element of tetromino should be shifted

## rotateTetromino(typeOfTetromino, rotationPhaseOfTetromino, tetrominoCoordinates)

Using this function, combine all the steps into final function:
Parameters:
- typeOfTetromino (same as above)
- rotationPhaseOfTetromino (same as above)
- tetrominoCoordinates - array of coordinate objects (x,y), which represents actual coordinates

Function should return new array of coordinates.
Resulting array should be sorted.
Function should use algorithm described at the top of the issue.

Write unit tests for function rotateTetromino for _every_ possible input of typeOfTetromino and rotationPhaseOfTetromino. Only one tetrominoCoordinates is needed per combination.
Examples:
- O-tetromino, one rotation
- T-tetromino, with "nose" pointing up, down, left, right
- I-tetromino, horizontal and vertical, one rotation each
- etc.

tetrominoCoordinates and expected result should be represented graphically

## incrementPhase(typeOfTetromino, currentRotationPhaseOfTetromino)

Because rotateTetromino accepts rotationPhaseOfTetromino externally, we need another helper function to update rotation phase.
Function accepts 2 parameters
- typeOfTetromino (same as above)
- currentRotationPhaseOfTetromino (same as above)

This function returns new rotation phase.

- If currentRotationPhaseOfTetromino is the last one, we should return return first phase possible
- If currentRotationPhaseOfTetromino is _not_ the last one, we return next rotation phase

Example:
typeOfTetromino - I
currentRotationPhaseOfTetromino - horizontal
return - vertical
Example 2:
typeOfTetromino - T
currentRotationPhaseOfTetromino - "nose" pointing to the right
return - "nose" pointing to the top

Cover all combinations of typeOfTetromino and currentRotationPhaseOfTetromino with unit tests.

### PR DESCRIPTION
**Pull request by PashaBeshchuk**

### CODE REVIEW

File: tetris.js [hunk 1]
```diff
@@ -308,29 +308,36 @@ module.exports = {
...
+	shiftCoordinates: function (arrayCoordinates, shift) {
 		let newCoordinates = []
-		for(let elem of arrayCoordinates){
+		for (let elem of arrayCoordinates) {
 			newCoordinates.push(this.shiftCoordinate(elem, shift))
 		}
 		return newCoordinates
+	},
+	determineShift: function (typeOfTetromino, rotationPhaseOfTetromino) {
+		return 2;
```
Comment (zbstof) [inline tetris.js]: Incorrect type returned

File: tetris.js [hunk 2]
```diff
@@ -308,29 +308,36 @@ module.exports = {
...
+	},
+	rotateTetromino: function (typeOfTetromino, rotationPhaseOfTetromino, tetrominoCoordinates) {
+		let pivot = this.calculatorCoordinateOfTopLeft(tetrominoCoordinates);
+		return this.rotateArrayOfCoordinates(tetrominoCoordinates, pivot)
```
Comment (zbstof) [inline tetris.js]: You didn't implement whole algorithm. Where is the call to determineShift and shiftCoordinates?

File: tetris.js [hunk 3]
```diff
@@ -333,11 +333,23 @@ module.exports = {
...
+		}else if(arrayCoordinatesOfTetromino === "O"){
+			return phases = ["stable"];
+		}
+	},
 	determineShift: function (typeOfTetromino, rotationPhaseOfTetromino) {
-		return 2;
+		return { x: 0, y: 0 };
 	},
 	rotateTetromino: function (typeOfTetromino, rotationPhaseOfTetromino, tetrominoCoordinates) {
+		let phase = determineShift(typeOfTetromino, rotationPhaseOfTetromino)
```
Comment (zbstof) [inline tetris.js]: Shift is not a phase

**Comment by zbstof** [comment]
Ok, please add tests now

File: test/test-rotateTetromino.js [hunk 4]
```diff
@@ -0,0 +1,375 @@
+var assert = require('chai').assert;
+var tetris = require('../tetris');
+describe("O-tetromino", function () {
+    it("O-tetromino rotation phase", function () {
+        let typeOfTetromino = "O";
+        let rotationPhaseOfTetromino = ["stable"];
```
Comment (zbstof) [inline test/test-rotateTetromino.js]: rotationPhaseOfTetromino should be one value, not a list of values.
How would rotateTetromino know in which phase is tetromino?
Please change in all tests

Comment (PashaBeshchuk) [reply test/test-rotateTetromino.js]: так все верно, тетрамино будет поворачивать в данных направлениях, а rotationPhaseOfTetromino указывает текущее положение

Comment (zbstof) [reply test/test-rotateTetromino.js]: 1. English, please.
2. You've already changed this, how is this relevant to my comment?

File: test/test-rotateTetromino.js [hunk 5]
```diff
@@ -0,0 +1,375 @@
+var assert = require('chai').assert;
+var tetris = require('../tetris');
+describe("O-tetromino", function () {
+    it("O-tetromino rotation phase", function () {
```
Comment (zbstof) [inline test/test-rotateTetromino.js]: All test names should reflect both initial and target phase, not only initial phase

File: test/test-rotateTetromino.js [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+            -XX--
+            -XX--
+            -----
+        `);
+        assert.deepEqual(resultFunction, expectedResult);
+    });
+});
+
+describe("J-tetromino", function () {
+    it("J-tetromino rotation , starts position down", function () {
```
[the remark this answers is not in the document]
Comment (zbstof) [reply test/test-rotateTetromino.js]: And other tests

**Comment by zbstof** [comment]
Shift is not a phase

File: test/test-rotateTetromino.js [hunk 7]
```diff
@@ -22,7 +22,7 @@ describe("O-tetromino", function () {
 });
 
 describe("J-tetromino", function () {
-    it("J-tetromino rotation right, starts position down", function () {
+    it("J-tetromino rotation dow to right", function () {
```
Comment (zbstof) [inline test/test-rotateTetromino.js]: down

File: test/test-rotateTetromino.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (PashaBeshchuk) [reply test/test-rotateTetromino.js]: 2. rotationPhaseOfTetromino - it is function witch return array coordinates or function witch return present state tetramino?

[the remark this answers is not in the document]
Comment (zbstof) [reply test/test-rotateTetromino.js]: this has nothing to do with my comment. Please stop commenting in unrelated places. 
If you have questions about requirements - please ask in the issue

File: tetris.js [hunk 8]
```diff
@@ -308,29 +308,48 @@ module.exports = {
...
-	shiftCoordinates: function(arrayCoordinates, shift){
+	shiftCoordinates: function (arrayCoordinates, shift) {
 		let newCoordinates = []
-		for(let elem of arrayCoordinates){
+		for (let elem of arrayCoordinates) {
 			newCoordinates.push(this.shiftCoordinate(elem, shift))
 		}
 		return newCoordinates
+	},
+	rotationPhaseOfTetromino: function (arrayCoordinatesOfTetromino){
```
Comment (zbstof) [inline tetris.js]: 1. this function is not used currently - please remove
2. please reformat this file

File: tetris.js [hunk 9]
```diff
@@ -308,29 +308,48 @@ module.exports = {
...
+		if(arrayCoordinatesOfTetromino === "L" || arrayCoordinatesOfTetromino === "J" || arrayCoordinatesOfTetromino === "T"){
+			return phases = ["left", "down", "right", "up"];
+		}else if(arrayCoordinatesOfTetromino === "I" || arrayCoordinatesOfTetromino === "S" || arrayCoordinatesOfTetromino === "Z"){
+			return phases = ["vertical", "horizontal"];
+		}else if(arrayCoordinatesOfTetromino === "O"){
+			return phases = ["stable"];
+		}
+	},
+	determineShift: function (typeOfTetromino, rotationPhaseOfTetromino) {
+		return { x: 0, y: 0 };
```
Comment (zbstof) [inline tetris.js]: Now, please implement this function so that all tests pass

File: tetris.js [hunk 10]
```diff
@@ -344,12 +344,61 @@ module.exports = {
 		}
 	},
 	determineShift: function (typeOfTetromino, rotationPhaseOfTetromino) {
-		return { x: 0, y: 0 };
+		if(rotationPhaseOfTetromino === "stable"){
```
Comment (zbstof) [inline tetris.js]: Please make sure all ifs have the following structure:
Outer if checks for tetromino type
Inner if checks for phase type

File: tetris.js [hunk 11]
```diff
@@ -308,29 +308,97 @@ module.exports = {
...
+			return phases = ["vertical", "horizontal"];
+		}else if(arrayCoordinatesOfTetromino === "O"){
+			return phases = ["stable"];
+		}
+	},
+	determineShift: function (typeOfTetromino, rotationPhaseOfTetromino) {
+		if(rotationPhaseOfTetromino === "stable"){
+			return { x:0, y:1 };
+		}
+		if(rotationPhaseOfTetromino === "up"){
```
Comment (zbstof) [inline tetris.js]: Please group all ifs by type first. You should have 7 outer ifs

File: tetris.js [hunk 12]
```diff
@@ -344,55 +344,21 @@ module.exports = {
 		}
 	},
 	determineShift: function (typeOfTetromino, rotationPhaseOfTetromino) {
-		if(rotationPhaseOfTetromino === "stable"){
+		if(rotationPhaseOfTetromino === "up" && typeOfTetromino === "J" || rotationPhaseOfTetromino === "up" && typeOfTetromino === "L"|| rotationPhaseOfTetromino === "down" && typeOfTetromino === "J" || rotationPhaseOfTetromino === "down" && typeOfTetromino === "L" || rotationPhaseOfTetromino === "stable"  || rotationPhaseOfTetromino === "vertical" && typeOfTetromino === "Z" || rotationPhaseOfTetromino === "vertical" && typeOfTetromino === "S") {
```
Comment (zbstof) [inline tetris.js]: There should be nested ifs
Please follow my previous comments

File: tetris.js [hunk 13]
```diff
@@ -333,31 +333,87 @@ module.exports = {
...
+			if (rotationPhaseOfTetromino === "up") {
+				return { x: 0, y: 1 };
+			}
+			if (rotationPhaseOfTetromino === "left") {
+				return { x: 0, y: 2 };
+			}
+			if (rotationPhaseOfTetromino === "down") {
+				return { x: 0, y: 1 };
+			}
+			if (typeOfTetromino === "J") {
```
Comment (zbstof) [inline tetris.js]: Mistake

File: tetris.js [hunk 14]
```diff
@@ -333,31 +333,87 @@ module.exports = {
 		}
 		return newCoordinates
 	},
-	rotationPhaseOfTetromino: function (arrayCoordinatesOfTetromino){
+	rotationPhaseOfTetromino: function (arrayCoordinatesOfTetromino) {
```
Comment (zbstof) [inline tetris.js]: You dont need this function, please remove

**Pull request merged**